src/iam-token.js

```diff
@@ -0,0 +1,146 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const axios = require('axios');
+const objectPath = require('object-path');
+const hash = require('object-hash');
+
+
+module.exports = class IamTokenGetter {
+  constructor() {
+    this.s3TokenCache = {};
+    this.log = require('./bunyan-api').createLogger('IamTokenGetter');
+  }
+  async fetchS3Token(iam, kubeResourceMeta, namespace) {
+    let apiKey;
+    const apiKeyAlpha1 = objectPath.get(iam, 'api_key');
+    const apiKeyStr = objectPath.get(iam, 'apiKey');
+    const apiKeyRef = objectPath.get(iam, 'apiKeyRef');
+
+    if (typeof apiKeyAlpha1 == 'string') {
+      apiKey = apiKeyAlpha1;
+    } else if (typeof apiKeyStr == 'string') {
+      apiKey = apiKeyStr;
+    } else if (typeof apiKeyAlpha1 == 'object') {
+      let secretName = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    } else if (typeof apiKeyRef == 'object') {
+      let secretName = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    }
+    if (!apiKey) {
+      return Promise.reject('Failed to find valid apikey to authenticate against iam');
+    }
+
+    let res = await this._requestToken(iam, apiKey);
+    return res.access_token;
+  }
+
+
+  async _requestToken(iam, apiKey) {
+    let token;
+
+    const apiKeyHash = hash(apiKey, { algorithm: 'shake256' });
+    // Check if the cache has the token. If it does have it, copy it into token
+    // for further processing
+    if (objectPath.has(this.s3TokenCache, [apiKeyHash])) {
+      // fetch cached token
+      token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+    }
+
+    // if we are able to fetch the token from the cache and it is not a Promise,
+    // then we need to check the expiry. If it is not expiring in the next 120 s
+    // return the token
+    if (!(token instanceof Promise)) {
+      if (token !== undefined) {
+        const expires = objectPath.get(token, 'expiration', 0); // expiration: time since epoch in seconds
+        // re-use cached token as long as we are more than 2 minutes away from expiring.
+        if (Date.now() < (expires - 120) * 1000) {
+          return token;
+        }
+      } else {
+        // if we get to this point the token is either undefined or it is stale data.
```

Comment:
Apparent logic error: if the token is not a promise but is expired or about to expire, it will not proceed to the `else` clause to make a new promise to get a new token.  Removing the `} else {` (and closing brace and reducing indentation) should correct this.

Reply:
Good catch. Thanks!